Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm

```diff
@@ -0,0 +1,357 @@
+{
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (element.isMultiRepresentationHEIC()) {
+        if (RefPtr image = element.image()) {
+            if (WebMultiRepresentationHEICAttachment *attachment = image->adapter().multiRepresentationHEIC()) {
+                RetainPtr attachmentString = adoptNS([[NSString alloc] initWithFormat:@"%C", static_cast<unichar>(NSAttachmentCharacter)]);
+                RetainPtr attributedString = adoptNS([[NSMutableAttributedString alloc] initWithString:attachmentString.get()]);
+                [attributedString addAttribute:WebMultiRepresentationHEICAttachmentAttributeName value:attachment range:NSMakeRange(0, 1)];
+                return attributedString;
+            }
+        }
+    }
+#endif
+
+    RetainPtr fileWrapper = fileWrapperForElement(element);
+    RetainPtr attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+    return [NSAttributedString attributedStringWithAttachment:attachment.get()];
+}
+
+static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const HTMLAttachmentElement& element)
+{
+    RetainPtr fileWrapper = fileWrapperForElement(element);
+    RetainPtr attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+    return [NSAttributedString attributedStringWithAttachment:attachment.get()];
+}
+
+static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor, HashMap<Element*, bool>& cache)
+{
+    if (!ancestor)
+        return false;
+
+    return cache.ensure(ancestor, [&] {
+        if (isMailBlockquote(*ancestor))
+            return true;
+
+        // FIXME: This should use the render style of the element instead.
+        if (RefPtr htmlElement = dynamicDowncast<HTMLElement>(ancestor); htmlElement && htmlElement->hasTagName(HTMLNames::preTag))
+            return true;
+
+        return hasAncestorQualifyingForWritingToolsPreservation(ancestor->parentElement(), cache);
+    }).iterator->value;
+}
+
+#pragma mark - EditingHTMLConverter definition.
+
+class EditingHTMLConverter {
+public:
+    explicit EditingHTMLConverter(const SimpleRange&, OptionSet<IncludedElement>);
+    ~EditingHTMLConverter();
+
+    AttributedString convert();
+
+private:
+    SimpleRange m_range;
+    OptionSet<IncludedElement> m_includedElements;
+};
+
+#pragma mark - EditingHTMLConverter implementation.
+
+EditingHTMLConverter::EditingHTMLConverter(const SimpleRange& range, OptionSet<IncludedElement> includedElements)
+    : m_range(range)
+    , m_includedElements(includedElements)
+{
+}
+
+EditingHTMLConverter::~EditingHTMLConverter() = default;
+
+// This function uses TextIterator, which makes offsets in its result compatible with HTML editing.
+AttributedString EditingHTMLConverter::convert()
+{
+#if PLATFORM(MAC)
+    auto fontManager = [NSFontManager sharedFontManager];
+#endif
+
+    HashMap<Element*, bool> m_elementQualifiesForWritingToolsPreservationCache;
```

Comment:
`WeakHashMap` or `RefPtr<Element>`.

Reply:
hm, I get this error when using WeakHashMap


```
OpenSource/WebKitBuild/Products/Debug/usr/local/include/wtf/WeakHashMap.h:372:16: error: no viable conversion from returned value of type 'WebCore::WeakPtrImplWithEventTargetData' to function return type 'RefType' (aka 'Ref<WTF::DefaultWeakPtrImpl>')
  372 |         return *key.weakPtrFactory().template createWeakPtr<T>(const_cast<T&>(key)).m_impl;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
OpenSource/WebKitBuild/Products/Debug/usr/local/include/wtf/WeakHashMap.h:201:36: note: in instantiation of function template specialization 'WTF::WeakHashMap<WebCore::Element, bool>::makeKeyImpl<WebCore::Element>' requested here
  201 |         auto result = m_map.ensure(makeKeyImpl(key), functor);
      |                                    ^
OpenSource/Source/WebCore/editing/cocoa/HTMLConverter.mm:2447:18: note: in instantiation of function template specialization 'WTF::WeakHashMap<WebCore::Element, bool>::ensure<(lambda at OpenSource/Source/WebCore/editing/cocoa/HTMLConverter.mm:2447:36)>' requested here
 2447 |     return cache.ensure(*ancestor, [&] {
 ```

Reply:
I assume you tried using `WeakHashMap<Element, bool, WeakPtrImplWithEventTargetData>>`?

Reply:
oops! I did end up getting it working, but then it looks like I accidentally lost that change when doing all the un-refactoring

Reply:
oh wait it is already in the diff